plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiSplitSource.java

```diff
@@ -0,0 +1,135 @@
+
+public class HudiSplitSource
+        implements ConnectorSplitSource
+{
+    private static final Logger log = Logger.get(HudiSplitSource.class);
+    private static final long IDLE_WAIT_TIME_MS = 10;
+    private final HoodieTableMetaClient metaClient;
+    private final boolean metadataEnabled;
+    private final boolean shouldSkipMetastoreForPartition;
+    private final HudiFileListing hudiFileListing;
+    private final ArrayDeque<ConnectorSplit> connectorSplitQueue;
+    private final HudiSplitBackgroundLoader splitLoader;
+    private final ScheduledExecutorService splitLoaderExecutorService;
+    private final ScheduledFuture splitLoaderFuture;
+
+    public HudiSplitSource(
+            ConnectorSession session,
+            HiveMetastore metastore,
+            Table table,
+            HudiTableHandle tableHandle,
+            Configuration conf,
+            Map<String, HiveColumnHandle> partitionColumnHandleMap)
+    {
+        this.metadataEnabled = isHudiMetadataEnabled(session);
+        this.shouldSkipMetastoreForPartition = shouldSkipMetaStoreForPartition(session);
+        this.metaClient = tableHandle.getMetaClient().orElseGet(() -> getMetaClient(conf, tableHandle.getBasePath()));
+        HoodieEngineContext engineContext = new HoodieLocalEngineContext(conf);
+        HoodieMetadataConfig metadataConfig = HoodieMetadataConfig.newBuilder()
+                .enable(metadataEnabled)
+                .build();
+        List<HiveColumnHandle> partitionColumnHandles = table.getPartitionColumns().stream()
+                .map(column -> partitionColumnHandleMap.get(column.getName())).collect(toList());
+        // TODO: fetch the query mode from config / query context
+        this.hudiFileListing = HudiFileListingFactory.get(HudiQueryMode.READ_OPTIMIZED,
+                metadataConfig, engineContext, tableHandle, metaClient, metastore, table,
+                partitionColumnHandles, shouldSkipMetastoreForPartition);
+        this.connectorSplitQueue = new ArrayDeque<>();
+        this.splitLoader = new HudiSplitBackgroundLoader(
+                session, tableHandle, metaClient, hudiFileListing, connectorSplitQueue);
+        this.splitLoaderExecutorService = Executors.newSingleThreadScheduledExecutor();
+        this.splitLoaderFuture = this.splitLoaderExecutorService.schedule(
+                this.splitLoader, 0, TimeUnit.MILLISECONDS);
+    }
+
+    @Override
+    public CompletableFuture<ConnectorSplitBatch> getNextBatch(ConnectorPartitionHandle partitionHandle, int maxSize)
+    {
+        if (isFinished()) {
+            return completedFuture(new ConnectorSplitBatch(new ArrayList<>(), true));
+        }
+
+        HoodieTimer timer = new HoodieTimer().startTimer();
+        List<ConnectorSplit> connectorSplits = new ArrayList<>();
+
+        while (!splitLoaderFuture.isDone() && connectorSplitQueue.isEmpty()) {
+            try {
+                Thread.sleep(IDLE_WAIT_TIME_MS);
```

Comment:
instead sleeping here you should return an uncompleted future and fill it in with data later on when you have enough splits available